Mir3 support

ptpt52 opened the pull request:
add a new subtarget named mt7620nand
new mt7620-nand driver
add a new device(Xiaomi Mi Router 3) base on this sub target

repost from https://github.com/lede-project/source/pull/1526

ptpt52 committed:
ramips: new MT7620 based boards with NAND flash

- add a new subtarget named mt7620nand
- new mt7620-nand driver

Signed-off-by: Chen Minqiang <ptpt52@gmail.com>

ptpt52 committed:
ramips: add Xiaomi Mi Router 3 support

Specification:
 - CPU: 580 MHz
 - Flash size: 128 MiB NAND
 - RAM size: 128 MiB DDR2
 - Wireless No1: SoC-integrated: MT7620A 2x2 MIMO 802.11b/g/n (2.4 GHz)
 - Wireless No2: On-board chip: MT7612E 2x2 MIMO 802.11a/n/ac (5 GHz)
 - Switch: MT7620 built-in 10/100 switch w/ vlan support
 - USB: 2.0 x1

Flash instructions:
 1. access ssh via https://wiki.openwrt.org/toh/xiaomi/mir3
 2. in ssh:
      nvram set flag_last_success=1
      nvram set boot_wait=on
      nvram set uart_en=1
      nvram commit
      mtd write lede-ramips-mt7620nand-miwifi-r3-squashfs-kernel1.bin kernel1
      mtd write lede-ramips-mt7620nand-miwifi-r3-squashfs-rootfs0.bin rootfs0
      reboot

If you want to revert back to stock fw while on LEDE:
  insert usb stick (FAT/FAT32) with stock fw renamed to miwifi.bin
  fw_setenv flag_last_success 0
  power off the device
  hold the reset button and power on the device, keep holding the reset button until yellow led starts blinking.

This is base on the new subtarget mt7620nand

Signed-off-by: Chen Minqiang <ptpt52@gmail.com>

ptpt52 commented:
I resolved the conflicts again, any help to get it merging?

EXSERENS commented:
finally supported! anybody help this merging?

nbd168 commented on target/linux/ramips/image/mt7620nand.mk:
```diff
@@ -8,7 +8,7 @@ define Device/miwifi-r3
   BLOCKSIZE := 128k
   PAGESIZE := 2048
-  KERNEL_SIZE := 1472k
+  KERNEL_SIZE := 1440k
```
Is this KERNEL_SIZE field based on a real limitation in the kernel partition size, or did you put in a hardcoded value based on whatever size your build produced?

ysc3839 replied:
Maybe there's a bad block in nand flash? Run `dmesg | grep nand`.

nbd168 commented on target/linux/ramips/mt7620nand/target.mk:
```diff
@@ -0,0 +1,15 @@
+FEATURES+=nand usb
+CPU_TYPE:=24kc
+
+DEFAULT_PACKAGES += kmod-rt2800-pci kmod-rt2800-soc kmod-mt76 wpad-mini
```
Here you should select kmod-mt76x2 instead of kmod-mt76. That way you leave out kmod-mt7603, which you don't need.

ptpt52 replied:
ok, Thank you for reminding

nbd168 commented on target/linux/ramips/mt7620nand/config-4.9:
```diff
@@ -203,10 +196,6 @@ CONFIG_SOC_MT7620=y
-CONFIG_SPI=y
-CONFIG_SPI_MASTER=y
-# CONFIG_SPI_MT7621 is not set
-CONFIG_SPI_RT2880=y
```
If you want to save some more space, you can disable jffs2 as well.

ptpt52 replied:
ok, for now It is likely that the kernel_size restriction can be removed
I will submit a new commit later...

mkresin commented on target/linux/ramips/base-files/lib/ramips.sh:
```diff
@@ -295,6 +295,9 @@ ramips_board_detect() {
 	*"MiWiFi Mini")
 		name="miwifi-mini"
 		;;
+	*"Xiaomi Mi Router R3")
```
Drop this line and use `xiaomi,miwifi-r3` as board name in all files.

mkresin commented on target/linux/ramips/dts/MIWIFI-R3.dts:
```diff
@@ -0,0 +1,180 @@
+	};
+};
+
+&gpio0 {
```
drop this node, gpio0 isn't used in your dts file.

mkresin commented on target/linux/ramips/dts/MIWIFI-R3.dts:
```diff
@@ -0,0 +1,180 @@
+	status = "okay";
+};
+
+&gpio2 {
```
drop this node, gpio0 isn't used in your dts file.

mkresin commented on target/linux/ramips/image/mt7620nand.mk:
```diff
@@ -0,0 +1,23 @@
+  IMAGE/rootfs0.bin := append-ubi | check-size $$$$(IMAGE_SIZE)
+  IMAGE/sysupgrade.tar := sysupgrade-tar | append-metadata
+  DEVICE_TITLE := Xiaomi Mi Router R3
+  SUPPORTED_DEVICES += miwifi-r3
```
you can drop the SUPPORTED_DEVICES

mkresin commented on target/linux/ramips/image/mt7620nand.mk:
```diff
@@ -0,0 +1,23 @@
+
+DEVICE_VARS +=
+
+define Device/miwifi-r3
```
`define Device/xiaomi_miwifi-r3` please

mkresin commented on target/linux/ramips/image/mt7620nand.mk:
```diff
@@ -0,0 +1,23 @@
+  SUPPORTED_DEVICES += miwifi-r3
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci uboot-envtools
+endef
+TARGET_DEVICES += miwifi-r3
```
`TARGET_DEVICES += xiaomi_miwifi-r3` please

ptpt52 reviewed (commented):
now it is based on a real limitation in the kernel partition size

mkresin commented on target/linux/ramips/dts/MIWIFI-R3.dts:
```diff
@@ -0,0 +1,180 @@
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "rgmii1";
```
drop the i2c group here. None of the pins of this group (`&gpio0 1`, `&gpio0 2`) is used in your dts.

mkresin commented on target/linux/ramips/dts/MIWIFI-R3.dts:
```diff
@@ -0,0 +1,180 @@
+		};
+	};
+
+	nand {
```
It is 

```
nand: nand@800 {
	reg = <0x800 0x100>;

	pinctrl-names = "default";
	pinctrl-0 = <&nand_pins>;
}
```

Furthermore the whole nand node should be moved to mt7620a.dtsi with `status = "disabled"` and only referenced here via

```
&nand {
	status = "okay";
};
```

The nand pinmux should/need to be added to mt7620a.dtsi as well:

```
nand_pins: nand {
	nand {
		ralink,group = "nd_sd";
		ralink,function = "nand";
	};
};
```

mkresin commented on target/linux/ramips/image/mt7620nand.mk:
```diff
@@ -0,0 +1,23 @@
+  SUPPORTED_DEVICES += miwifi-r3
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci uboot-envtools kmod-mt76x2
+endef
+TARGET_DEVICES += miwifi-r3
```
`TARGET_DEVICES += xiaomi_miwifi-r3` please

mkresin commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -28,6 +28,7 @@ all0256n-8M|\
 all5002)
 	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x10000" "0x10000"
 	;;
+miwifi-r3|\
```
Any reason why you added u-boot env support for this board? As far as I can see, there is nothing which requires write access to the u-boot env.

ptpt52 replied:
it is need to revert back to stock firmware.
fw_setenv flag_last_success 0

mkresin reviewed (changes requested):
Please make sure that it works with the changes requested before updating the PR. Might be that I requested something that doesn't work.

Please squash the fixes of "ramips: add Xiaomi Mi Router 3 support" into this commit. It doesn't make sense to keep them as extra commits.

9000000 commented:
Please fix error

Sat Mar 17 00:46:02 2018 
daemon.err hostapd: Configuration file: /var/run/hostapd-phy0.conf
Sat Mar 17 00:46:03 2018
daemon.err hostapd: Configuration file: /var/run/hostapd-phy1.conf
Sat 
Sat Mar 17 07:54:51 2018 
daemon.notice hostapd: ACS: Survey is missing noise floor
Sat Mar 17 07:54:51 2018 
daemon.notice hostapd: ACS: Survey is missing noise floor
Sat Mar 17 07:54:51 2018 
daemon.notice hostapd: ACS: Survey is missing noise floor
Sat Mar 17 07:58:21 2018 
daemon.notice hostapd: 20/40 MHz operation not permitted on channel pri=11 sec=7 based on overlapping BSSes
Sat Mar 17 07:58:21 2018 
daemon.err hostapd: Using interface wlan1 with hwaddr 28:6c:07:4c:42:37 and ssid "Quick_Snow"

ptpt52 commented:
`
nand: nand@800 {
	reg = <0x800 0x100>;

	pinctrl-names = "default";
	pinctrl-0 = <&nand_pins>;

	...
}
`
@mkresin 
how to make it work? seems no driver for nand
I set it to compatible = "mediatek,mt2701-nfc";
no work.

ptpt52 commented:
`
Starting kernel ...

[    0.000000] Linux version 4.14.25 (terminator@SC-GAME) (gcc version 7.3.0 (OpenWrt GCC 7.3.0 r6427+43-ed3860c)) #0 Sat Mar 17 01:03:59 2018
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7620A ver:2 eco:6
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019650 (MIPS 24KEc)
[    0.000000] MIPS: machine is Xiaomi Mi Router R3
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 08000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000007ffffff]
[    0.000000] random: get_random_bytes called from start_kernel+0x8c/0x474 with crng_init=0
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 32512
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] PID hash table entries: 512 (order: -1, 2048 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 4, 65536 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Writing ErrCtl register=000344d4
[    0.000000] Readback ErrCtl register=000344d4
[    0.000000] Memory: 124368K/131072K available (3957K kernel code, 175K rwdata, 896K rodata, 212K init, 221K bss, 6704K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] CPU Clock: 580MHz
[    0.000000] clocksource: systick: mask: 0xffff max_cycles: 0xffff, max_idle_ns: 583261500 ns
[    0.000000] systick: enable autosleep mode
[    0.000000] systick: running - mult: 214748, shift: 32
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000012] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.007588] Calibrating delay loop... 385.84 BogoMIPS (lpj=1929216)
[    0.073558] pid_max: default: 32768 minimum: 301
[    0.078263] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.084641] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.093655] devtmpfs: initialized
[    0.101065] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.110608] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.116582] pinctrl core: initialized pinctrl subsystem
[    0.122175] NET: Registered protocol family 16
[    0.388198] PCI host bridge /pcie@10140000 ranges:
[    0.392814]  MEM 0x0000000020000000..0x000000002fffffff
[    0.397862]   IO 0x0000000010160000..0x000000001016ffff
[    0.410589] random: fast init done
[    0.420649] rt2880_gpio 10000600.gpio: registering 24 gpios
[    0.426107] rt2880_gpio 10000600.gpio: registering 24 irq handlers
[    0.432555] rt2880_gpio 10000638.gpio: registering 16 gpios
[    0.437992] rt2880_gpio 10000638.gpio: registering 16 irq handlers
[    0.444947] rt2880_gpio 10000660.gpio: registering 32 gpios
[    0.450397] rt2880_gpio 10000660.gpio: registering 32 irq handlers
[    0.457865] PCI host bridge to bus 0000:00
[    0.461858] pci_bus 0000:00: root bus resource [mem 0x20000000-0x2fffffff]
[    0.468472] pci_bus 0000:00: root bus resource [io  0xffffffff]
[    0.474232] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.480815] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.488979] pci 0000:00:00.0: bridge configuration invalid ([bus 00-00]), reconfiguring
[    0.497425] pci 0000:00:00.0: BAR 0: no space for [mem size 0x80000000]
[    0.503837] pci 0000:00:00.0: BAR 0: failed to assign [mem size 0x80000000]
[    0.510553] pci 0000:00:00.0: BAR 8: assigned [mem 0x20000000-0x200fffff]
[    0.517157] pci 0000:00:00.0: BAR 9: assigned [mem 0x20100000-0x201fffff pref]
[    0.524159] pci 0000:00:00.0: BAR 1: assigned [mem 0x20200000-0x2020ffff]
[    0.530776] pci 0000:01:00.0: BAR 0: assigned [mem 0x20000000-0x200fffff 64bit]
[    0.537866] pci 0000:01:00.0: BAR 6: assigned [mem 0x20100000-0x2010ffff pref]
[    0.544883] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.549689] pci 0000:00:00.0:   bridge window [mem 0x20000000-0x200fffff]
[    0.556291] pci 0000:00:00.0:   bridge window [mem 0x20100000-0x201fffff pref]
[    0.566571] clocksource: Switched to clocksource systick
[    0.575093] NET: Registered protocol family 2
[    0.580241] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.587020] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.593142] TCP: Hash tables configured (established 1024 bind 1024)
[    0.599472] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.605097] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.611491] NET: Registered protocol family 1
[    0.617538] rt-timer 10000100.timer: maximum frequency is 1220Hz
[    0.624138] Crashlog allocated RAM at address 0x3f00000
[    0.630600] workingset: timestamp_bits=30 max_order=15 bucket_order=0
[    0.642469] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.648121] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.666829] io scheduler noop registered
[    0.670575] io scheduler deadline registered (default)
[    0.675890] ralink-usb-phy usbphy: failed to get sysctl registers
[    0.682964] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.691399] console [ttyS0] disabled
[    0.694849] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 20, base_baud = 2500000) is a Palmchip BK-3103
[    0.704540] console [ttyS0] enabled
[    0.704540] console [ttyS0] enabled
[    0.711568] bootconsole [early0] disabled
[    0.711568] bootconsole [early0] disabled
[    0.720945] cacheinfo: Failed to find cpu0 device node
[    0.726198] cacheinfo: Unable to detect cache hierarchy for CPU 0
[    0.733006] rt2880-pinmux pinctrl: pin io45 already requested by pinctrl; cannot claim for 10000800.nand
[    0.742751] rt2880-pinmux pinctrl: pin-45 (10000800.nand) status -22
[    0.749251] rt2880-pinmux pinctrl: could not request pin 45 (io45) from group nd_sd  on device rt2880-pinmux
[    0.759289] mtk-nand 10000800.nand: Error applying setting, reverse things back
[    0.766781] mtk-nand: probe of 10000800.nand failed with error -22
[    0.773805] libphy: Fixed MDIO Bus: probed
[    0.782489] gsw: setting port4 to ephy mode
[    0.786847] mtk_soc_eth 10100000.ethernet eth0 (uninitialized): port 1 link up (100Mbps/Full duplex)
[    0.796205] mtk_soc_eth 10100000.ethernet: generated random MAC address 5a:9d:1f:3b:77:26
[    0.804645] mtk_soc_eth 10100000.ethernet: loaded mt7620 driver
[    0.811399] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    0.821313] rt2880_wdt 10000120.watchdog: Initialized
[    0.827955] NET: Registered protocol family 10
[    0.835502] Segment Routing with IPv6
[    0.839436] NET: Registered protocol family 17
[    0.844025] bridge: filtering via arp/ip/ip6tables is no longer available by default. Update your scripts to load br_netfilter if you need this.
[    0.857264] 8021q: 802.1Q VLAN Support v1.8
[    0.866935] VFS: Cannot open root device "(null)" or unknown-block(0,0): error -6
[    0.874578] Please append a correct "root=" boot option; here are the available partitions:
[    0.883163] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(0,0)
[    0.892827] Rebooting in 1 seconds..

`

I post the boot log here.

9000000 commented:
new firmware image not update

Error: The uploaded image file does not contain a supported format. Make sure that you choose the generic image format for your platform

ptpt52 commented:
@9000000 
hi, new version you need to upgrade via

> sysupgrade -F /tmp/......sysupgrade.tar

9000000 commented:
@ptpt52 
Please fix error

18 08:27:16 2018 daemon.notice hostapd: handle_probe_req: send failed
Sun Mar 18 08:27:16 2018 daemon.notice hostapd: handle_probe_req: send failed
Sun Mar 18 08:27:16 2018 daemon.notice hostapd: 20/40 MHz operation not permitted on channel pri=11 sec=7 based on overlapping BSSes
Sun Mar 18 08:27:16 2018 daemon.err hostapd: Using interface wlan1 with hwaddr 28:6c:07:4c:42:37 and ssid "Quick_Snow"

9000000 commented:
New fw 
Still error
Sun Mar 18 08:32:50 2018 kern.err kernel: [  211.104526] ieee80211 phy1: rt2x00queue_write_tx_frame: Error - Dropping frame due to full tx queue 2
Su

ptpt52 commented:
see https://github.com/openwrt/openwrt/pull/803

> Still error
Sun Mar 18 08:32:50 2018 kern.err kernel: [ 211.104526] ieee80211 phy1: rt2x00queue_write_tx_frame: Error - Dropping frame due to full tx queue 2

the error printk msg still there. but it does not stuck wifi. that is ok.

9000000 commented:
You can fix this

`Fri Mar 23 11:58:29 2018 user.warn igmpproxy[2108]: MRT_DEL_MFC; Errno(2): No such file or directory
Fri Mar 23 12:02:39 2018 user.warn igmpproxy[2108]: MRT_DEL_MFC; Errno(2): No such file or directory
Fri Mar 23 12:06:50 2018 user.warn igmpproxy[2108]: MRT_DEL_MFC; Errno(2): No such file or directory
Fri Mar 23 12:10:59 2018 user.warn igmpproxy[2108]: MRT_DEL_MFC; Errno(2): No such file or directory
Fri Mar 23 12:15:`

9000000 commented:
Please update kernel  3.14.34
https://downloads.openwrt.org/snapshots/targets/ramips/mt7620/packages/kernel_4.14.34-1-8ec7184a251df65efc47cc3fdacf99d4_mipsel_24kc.ipk

mkresin commented:
@ysc3839 My main concern is the mt7620 nand driver. It looks like copy/paste of the already ugly ralink driver. Beside the code quality, I'm concerned about the licence of the ralink driver. Is it GPL licenced so that we can use/modify the ralink driver? At least within the PR there is no licence header, copyright or similar.

The best would be, if the driver goes upstream (linux kernel) or at least the mt7621 nand driver is reused. It might be that the upstream mediatek ARM nand driver can be extended to support the mt7620 nand.

Some hours ago @nbd168 committed f95a4aa5658e8686831e88a530cbd596741ff573. I kind of lost the overview about the Mir boards, since they are all named very similar and I fail to see which one f95a4aa5658e8686831e88a530cbd596741ff573 targets. In best case the mt7620-nand driver from this PR isn't required any longer. 

I'm not going to take care of this PR, as @blogic picked the [original PR](https://github.com/lede-project/source/pull/1526)  (and hence got this one assigned). It can be assumed that he knows the hardware way better than I do.

ptpt52 commented:
> In best case the mt7620-nand driver from this PR isn't required any longer.

I am trying to do reusing the mt7621 nand driver but fail for my limited understanding.
I hope someone can do this. I prefer to use the upstream mt7621 nand driver.

ysc3839 commented on target/linux/ramips/patches-4.14/0038-mtd-ralink-add-mt7620-nand-driver.patch:
```diff
@@ -0,0 +1,2385 @@
++
++module_platform_driver(mtk_nand_driver);
++
++MODULE_LICENSE("GPL");
```
@mkresin The module is GPL licensed. This patch exist in `lede-17.01` branch [link](https://github.com/openwrt/openwrt/blob/lede-17.01/target/linux/ramips/patches-4.4/0038-mtd-ralink-add-mt7620-nand-driver.patch). But removed in master.

mkresin replied:
So it got lost/was removed intentional with the [update to kernel 4.9](https://git.openwrt.org/fddc78bc116bc043fb5cd92dbc45cfd054c21af8). That is something I expected to be mentioned in the commit message instead of "new mt7620-nand driver".

It's quite amazing that people are curious why their PRs stuck... Anyway, as said before I'm not the right one for this PR.

aparcar commented:
My mir3 is all dusty, is this PR still alive?

ptpt52 commented:
@mkresin please re-review thx.

blogic commented:
this PR adds a new subtarget for an out of tree driver that wont ever go upstream and is of abysmal quality. further this driver will break heavily when we move on to the next kernel release. there is no one that will have time to maintain this driver. sorry ....

dmitrygribenchuk commented:
Hello. Is support it added now?

Neustradamus commented:
Any news about the support of Xiaomi Mi Router 3?

There were 2 PRs:
- https://github.com/openwrt/openwrt/pull/597
- https://github.com/openwrt/openwrt/pull/764
